### PR TITLE
`filter` block: return all the data with both Result.SUCCESS and Result.FILTERED results #173

### PR DESCRIPTION
**Pull request by zalmane**

### CODE REVIEW

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -21,7 +21,9 @@ jobs:
       - name: Install dependencies
         run: |
           cd core
-          pip install . 
+          pip install wheel
+          CASS_DRIVER_NO_CYTHON=1 pip install cassandra-driver
+          pip install .
```
Comment (spicy-sauce) [inline .github/workflows/unit-tests.yml:26]: why is it needed here too?

File: core/src/datayoga_core/blocks/add_field/block.py [hunk 2]
```diff
@@ -22,18 +22,20 @@ def init(self, context: Context = None):
...
         logger.debug(f"Running {self.get_block_name()}")
-        for row in data:
-            for field in self.fields:
-                obj = row
-                field_path = utils.split_field(field)
+        for field in self.fields:
+            expression_results = self.fields[field].search_bulk(data)
+            field_path = utils.split_field(field)
 
+            for i,row in enumerate(data):
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/add_field/block.py]: add space after comma (I wonder why it doesn't format it for you this way automatically).

File: core/src/datayoga_core/blocks/filter/block.py [hunk 3]
```diff
@@ -16,6 +16,6 @@ def init(self, context: Context = None):
 
     async def run(self, data: List[Dict[str, Any]]) -> Tuple[List[Dict[str, Any]], List[Result]]:
         logger.debug(f"Running {self.get_block_name()}")
-        return_data = self.expression.filter(data)
-
-        return utils.all_success(return_data)
+        return_data = self.expression.filter(data,tombstone=True)
+        # mark filtered rows
+        return [x for x in return_data if x is not None],[Result(Status.FILTERED) if x is None else Result(Status.SUCCESS) for x in return_data]
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/filter/block.py]: add space after comma (I wonder why it doesn't format it for you this way automatically).

File: core/src/datayoga_core/blocks/map/block.py [hunk 4]
```diff
@@ -17,19 +17,22 @@ def init(self, context: Context = None):
...
-        return utils.all_success(return_data)
+        mapped_rows = self.expression.search_bulk(data)
+        # we always add the internal fields back, in case they were dropped in the map operation
+        for mapped_row,original_row in zip(mapped_rows,data):
+            original_msg_id= original_row.get(Block.MSG_ID_FIELD)
+            original_opcode= original_row.get(Block.OPCODE_FIELD)
+            if original_msg_id is not None:
+                mapped_row[Block.MSG_ID_FIELD] = original_msg_id
+            if original_opcode is not None:
+                mapped_row[Block.OPCODE_FIELD] = original_opcode
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/map/block.py:36]: maybe we should consider adding those fields by their prefix, so we won't need to explicitly list them one by one...

File: core/src/datayoga_core/blocks/map/block.py [hunk 5]
```diff
@@ -17,19 +17,22 @@ def init(self, context: Context = None):
...
+
+        return utils.all_success(mapped_rows)
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/map/block.py:38]: I think that when you try to map a missing field, you'll get an error, so in this case we either should nullify this field or reject the whole record.
I now see in the tests that in this case the field will be set to None, if this is indeed the case, please close.

File: core/src/datayoga_core/expression.py [hunk 6]
```diff
@@ -44,8 +47,9 @@ def compile(self, expression: str):
         """
         pass
```
Comment (spicy-sauce) [inline core/src/datayoga_core/expression.py]: you should replace `pass` with:
`raise NotImplementedError`

File: core/src/datayoga_core/expression.py [hunk 7]
```diff
@@ -55,35 +59,48 @@ def search(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
```
Comment (spicy-sauce) [inline core/src/datayoga_core/expression.py]: you should replace pass with:
raise NotImplementedError

File: core/src/datayoga_core/expression.py [hunk 8]
```diff
@@ -55,35 +59,48 @@ def search(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
-        # we turn off check_same_thread to gain performance benefit by reusing the same connection object. safe to use since we are only creating in memory structures
-        self.conn = sqlite3.connect(":memory:", check_same_thread=False)
-        self.expression = expression
+        Args:
+            data (List[Dict[str, Any]]): Data
+
+        Returns:
+            List[Dict[str, Any]]: Transformed data
+        """
+        pass
```
Comment (spicy-sauce) [inline core/src/datayoga_core/expression.py]: you should replace pass with:
raise NotImplementedError

File: core/src/datayoga_core/expression.py [hunk 9]
```diff
@@ -55,35 +59,48 @@ def search(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
 
-    def filter(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    def filter(self, data: List[Dict[str, Any]],tombstone:bool = False) -> List[Union[Dict[str, Any],None]]:
```
Comment (spicy-sauce) [inline core/src/datayoga_core/expression.py]: add space after comma (I wonder why it doesn't format it for you this way automatically).

File: core/src/datayoga_core/expression.py [hunk 10]
```diff
@@ -55,35 +59,48 @@ def search(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
...
+        # filter is essentially using a single field returning 0 or 1 for the condition
+        if tombstone:
+            return [row[0] if row[1] else None for row in zip(data,self.search_bulk(data))]
+        else:
+            return [row[0] for row in zip(data,self.search_bulk(data)) if row[1]]
 
-        column_names = data_inner[0].keys()
+class SQLExpression(Expression):
+    def compile(self, expression: str):
+        # we turn off check_same_thread to gain performance benefit by reusing the same connection object. safe to use since we are only creating in memory structures
```
Comment (spicy-sauce) [inline core/src/datayoga_core/expression.py]: split comment to two lines.

File: core/src/datayoga_core/expression.py [hunk 11]
```diff
@@ -98,18 +115,19 @@ def _get_cte(self, data: List[Any]) -> str:
...
+            return [x.get("expr") for x in results]
+        else:
+            return results
+
+    def search(self, data: Dict[str, Any]) -> Any:
+        return self.search_bulk([data])[0]
+
 
-    def exec_sql(self, data: List[Dict[str, Any]], expression: str) -> List[Dict[str, Any]]:
+    def exec_sql(self, data: List[Dict[str, Any]], expressions: Dict[str,str]) -> List[Dict[str, Any]]:
```
Comment (spicy-sauce) [inline core/src/datayoga_core/expression.py]: str,str --> str, str

File: core/src/datayoga_core/step.py [hunk 12]
```diff
@@ -52,7 +53,7 @@ async def process(self, messages: List[Dict[str, Any]]):
         if not self.initialized:
             await self.start_pool()
             self.initialized = True
-
+        logger.warning(f"got {messages}")
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step.py]: why is it a warning message?

File: core/src/datayoga_core/step.py [hunk 13]
```diff
@@ -63,15 +64,21 @@ async def run(self, worker_id: int):
...
-                if self.next_step:
-                    # process downstream
-                    await self.next_step.process(processed_entries)
-                else:
-                    # we are a last channel, propagate the ack upstream
-                    # TODO: verify that all entries have a msg id otherwise raise consistency error
-                    self.done([x[Block.MSG_ID_FIELD] for x in processed_entries], results)
+                # handle filtered. anything not processed or rejected
+                filtered_entries = [payload_with_result[0][Block.MSG_ID_FIELD] for payload_with_result in zip(entry,results) if payload_with_result[1].status == Status.FILTERED]
+                logger.debug(f"filtered entries : {filtered_entries} processed entries: {processed_entries}")
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step.py]: logger.debug(f"filtered entries: {filtered_entries}, processed entries: {processed_entries}")

File: core/src/datayoga_core/step.py [hunk 14]
```diff
@@ -63,15 +64,21 @@ async def run(self, worker_id: int):
...
+                if len(filtered_entries)>0:
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step.py]: if len(filtered_entries) > 0:

File: core/src/datayoga_core/step.py [hunk 15]
```diff
@@ -63,15 +64,21 @@ async def run(self, worker_id: int):
...
+                    # ack the filtered entries
+                    self.done(filtered_entries, [Result(Status.FILTERED)] * len(filtered_entries))
+                if len(processed_entries)>0:
```
Comment (spicy-sauce) [inline core/src/datayoga_core/step.py]: if len(processed_entries) > 0:

File: core/tests/test_jmespath_performance.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
+import logging
+import time
+
+import datayoga_core as dy
```
Comment (spicy-sauce) [inline core/tests/test_jmespath_performance.py]: not in use

File: core/tests/test_sql_performance.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
+import logging
+import time
+
+import datayoga_core as dy
```
Comment (spicy-sauce) [inline core/tests/test_sql_performance.py]: not in use

File: integration-tests/test_redis_to_cassandra.py [hunk 18]
```diff
@@ -10,7 +12,9 @@
 TABLE = f"{KEYSPACE}.emp"
 
 
-def test_redis_to_cassandra():
+@pytest.fixture(scope="module")
+def prepare_db(request):
```
Comment (spicy-sauce) [inline integration-tests/test_redis_to_cassandra.py]: can we omit `request`? if not, can you add its type?

File: integration-tests/test_redis_to_cassandra.py [hunk 19]
```diff
@@ -27,12 +31,24 @@ def test_redis_to_cassandra():
 
     run_job("tests.redis_to_cassandra")
 
+    yield
+
+    # cleanup
+    redis_container.stop()
+    cassandra_container.stop()
+
+def test_total_records(prepare_db):
```
Comment (spicy-sauce) [inline integration-tests/test_redis_to_cassandra.py:39]: can you add the `prepare_db` type?

File: integration-tests/test_redis_to_cassandra.py [hunk 20]
```diff
@@ -27,12 +31,24 @@ def test_redis_to_cassandra():
...
+    session = cassandra_utils.get_cassandra_session(["localhost"])
     total_employees = session.execute(f"select count(*) as total from {TABLE}").one()
     assert total_employees.total == 3
 
+def test_filtered_record(prepare_db):
```
Comment (spicy-sauce) [inline integration-tests/test_redis_to_cassandra.py:44]: can you add the `prepare_db` type?

File: integration-tests/test_redis_to_cassandra.py [hunk 21]
```diff
@@ -27,12 +31,24 @@ def test_redis_to_cassandra():
...
+    session = cassandra_utils.get_cassandra_session(["localhost"])
     first_employee = session.execute(f"select * from {TABLE} where id = 1").one()
     assert first_employee is None
 
+def test_records(prepare_db):
```
Comment (spicy-sauce) [inline integration-tests/test_redis_to_cassandra.py:49]: can you add the `prepare_db` type?

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: .github/workflows/unit-tests.yml (hunk 1 above)
Comment (zalmane) [reply .github/workflows/unit-tests.yml:26]: separate container. speeds up the install.

File: core/src/datayoga_core/blocks/add_field/block.py (hunk 2 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/add_field/block.py]: done

File: core/src/datayoga_core/blocks/filter/block.py (hunk 3 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/filter/block.py]: done

File: core/src/datayoga_core/blocks/map/block.py (hunk 4 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/map/block.py:36]: it was like that originally, but then we need to scan each and every field which is a performance drag. that is when I saw the other field of the 'result_code'

File: core/src/datayoga_core/expression.py (hunk 11 above)
Comment (zalmane) [reply core/src/datayoga_core/expression.py]: done

File: core/src/datayoga_core/step.py (hunk 12 above)
Comment (zalmane) [reply core/src/datayoga_core/step.py]: debug. removed

File: core/src/datayoga_core/step.py (hunk 13 above)
Comment (zalmane) [reply core/src/datayoga_core/step.py]: done

File: core/src/datayoga_core/step.py (hunk 14 above)
Comment (zalmane) [reply core/src/datayoga_core/step.py]: done

File: core/src/datayoga_core/step.py (hunk 15 above)
Comment (zalmane) [reply core/src/datayoga_core/step.py]: done

File: core/tests/test_jmespath_performance.py (hunk 16 above)
Comment (zalmane) [reply core/tests/test_jmespath_performance.py]: done

File: core/tests/test_sql_performance.py (hunk 17 above)
Comment (zalmane) [reply core/tests/test_sql_performance.py]: done

File: integration-tests/test_redis_to_cassandra.py (hunk 18 above)
Comment (zalmane) [reply integration-tests/test_redis_to_cassandra.py]: removed

File: core/src/datayoga_core/blocks/map/block.py (hunk 4 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/map/block.py:36]: I see, OK, we will take care of the `result_code` separately.

File: core/src/datayoga_core/expression.py (hunk 6 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/expression.py]: fixed

File: core/src/datayoga_core/expression.py (hunk 7 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/expression.py]: fixed

File: core/src/datayoga_core/expression.py (hunk 8 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/expression.py]: done

**Review by spicy-sauce — APPROVED** [review]

**Pull request merged**